test/jit/test_class_type.py

```diff
@@ -1308,3 +1308,36 @@ def fn() -> bool:
 
         with self.assertRaisesRegexWithHighlight(RuntimeError, r"Class does not define __delitem__", "example[key]"):
             self.checkScript(fn, ())
+
+    def test_recursive_script_builtin_type_resolution(self):
+        """
+        Test resolution of built-in torch types(e.g. torch.Tensor, torch.device) when a class is recursively compiled.
+        """
+        # A will be implicitly compiled because it is not annotated with @torch.jit.script
+        # but is used in g() below.
+        tensor_t = torch.Tensor
+        device_t = torch.device
+
+        class A(object):
+            def __init__(self):
+                pass
+
+            def f(self, x: tensor_t, y: torch.device):
+                return x.to(device=y)
+
+            def h(self, x: device_t) -> device_t:
+                return x
+
+            def i(self) -> 'A':
+                return A()
+
+            def j(self, a: List[int]):
```

Comment:
Could you add a test case where a return type has a contained type, like `List[device_t]` ? if that doesnt work it can be filed as a separate issue as well, this is already a pretty big PR